cupy/core/core.pyx

```diff
@@ -2890,6 +2893,151 @@ cpdef ndarray tensordot_core(
     return ret
 
 
+cdef struct cuComplex:
+    float x, y
+
+
+cdef struct cuDoubleComplex:
+    double x, y
+
+
+cpdef ndarray tensordot_core_v11(
```

Comment:
Yes, code duplication is a concern of mine as well :sweat:
Since cublasGemmEx allows you to select a different data type for output matrix C than the data type of input matrices A and B, I was thinking of using this to reduce amount of copy after gemm (this was not implemented yet). That's why I was branching out early, but there aren't that many opportunities for copy reduction, so I'm going to prioritize reducing code duplication first.